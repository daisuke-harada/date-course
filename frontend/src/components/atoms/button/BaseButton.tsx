import { memo, ReactNode, VFC } from "react";

type Props = {
  children: ReactNode
  addClassNames: string
}
export const BaseButton: VFC<Props> = memo((props) => {
  const {children, addClassNames} = props;
  return(
   <button className={`btn btn-salmon ${addClassNames}`}>{children}</button>
  );
});