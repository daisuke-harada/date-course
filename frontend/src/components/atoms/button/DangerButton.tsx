import { memo, ReactNode, VFC } from "react";
import tw from "tailwind-styled-components";

type Props = {
  children: ReactNode;
  onClickEvent?: React.MouseEventHandler<HTMLButtonElement>,
  dataE2e?: string,
};

const Button = tw.button`btn btn-red w-full`;

export const DangerButton: VFC<Props> = memo((props) => {
  const {children, onClickEvent, dataE2e} = props;
  return(
    <Button data-e2e={dataE2e} onClick={onClickEvent}>{children}</Button>
  );
});