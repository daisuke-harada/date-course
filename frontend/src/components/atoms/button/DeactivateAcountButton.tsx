import { memo, VFC } from "react";
import tw from 'tailwind-styled-components';

import { useDeactivateAccountButtonAction } from "hooks/useDeactivateAccountButtonAction";

const ButtonParentDiv = tw.div`text-center p-1 mx-6 my-4`;

export const DeactivateAcountButton: VFC = memo(() => {
  const { onCLickDeactivateAccountAction, loginStatus }= useDeactivateAccountButtonAction();

  return(
    <>
      {
        loginStatus.status === true &&
        <ButtonParentDiv>
          <button className='text-red-600 hover:text-red-800 font-bold mb-5' onClick={onCLickDeactivateAccountAction}>退会</button>
        </ButtonParentDiv>
      }
    </>
  );

});