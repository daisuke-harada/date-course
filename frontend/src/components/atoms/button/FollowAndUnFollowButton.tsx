import { client } from 'lib/api/client';
import { memo, useEffect, useState, VFC } from 'react';
import { useRecoilState, useRecoilValue } from 'recoil';
import { currentUserState, loginStatusState } from 'store/session';
import tw from 'tailwind-styled-components';
import { UserResponseData } from 'types/users/response';

type Props = {
  dataE2e?: string,
  userId: number,
  addClassName?: string,
  setUsers?: React.Dispatch<React.SetStateAction<UserResponseData[]>>,
  setUser?: React.Dispatch<React.SetStateAction<UserResponseData>>
};

const FollowButton = tw.button`btn btn-yellow-green text-sm`
const UnfollowButton = tw.button`btn btn-unfollow`;

export const FollowAndUnFollowButton: VFC<Props> = memo((props) => {
  const {dataE2e, userId, setUsers, setUser, addClassName} = props;

  const [currentUser, setCurrentUser] = useRecoilState(currentUserState);
  const getLoginStatus = useRecoilValue(loginStatusState);
  const [currentUserId, setCurrentUserId] = useState<number>(0);

  useEffect(() => {
    currentUser.user &&
    setCurrentUserId(currentUser.user.id);
  },[currentUser, userId]);

  const onClickFollowAction = () => {
    client.post('relationships', {
      currentUserId: currentUser.user.id,
      followedUserId: userId,
    }).then(response => {
      response.data.status === 'created' && setUsers && setUsers(response.data.users);
      response.data.status === 'created' && setCurrentUser({user: response.data.currentUser});
      response.data.status === 'created' && setUser && setUser(response.data.followedUser);
    });
  };

  const onClickUnfollowAction = () => {
    client.delete(`relationships/${currentUserId}/${userId}`).then(response => {
      response.data.status === 'deleted' && setUsers && setUsers(response.data.users);
      response.data.status === 'deleted' && setUser && setUser(response.data.unfollowedUser);
      response.data.status === 'deleted' && setCurrentUser({user: response.data.currentUser});
    });
  };


  return(
    <>
      {
        getLoginStatus.status === true &&
        currentUserId !== userId &&
        (currentUser.user.followingIds && currentUser.user.followingIds.includes(userId)?
        <UnfollowButton data-e2e={dataE2e} className={addClassName} onClick={onClickUnfollowAction}>フォロー中</UnfollowButton>
        :
        <FollowButton data-e2e={dataE2e} className={addClassName} onClick={onClickFollowAction}>フォロー</FollowButton>
        )
      }
    </>
  );
});