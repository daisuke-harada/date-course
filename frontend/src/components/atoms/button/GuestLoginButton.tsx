import { memo, ReactNode, VFC } from "react";

type Props = {
  children: ReactNode;
  buttonSize: string;
};

export const GuestLoginButton: VFC<Props> = memo((props) => {
  const {children, buttonSize} = props;
  return(
   <button className={`btn btn-yellow-green ${buttonSize}`}>{children}</button>
  );
});