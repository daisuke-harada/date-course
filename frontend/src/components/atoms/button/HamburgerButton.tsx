import { memo, VFC } from "react";

type Props = {
  onClickNavBarSwitch: React.MouseEventHandler<HTMLButtonElement> | undefined;
  isOpen: boolean;
};

export const HamburgerButton: VFC<Props> = memo((props) => {
  const { onClickNavBarSwitch, isOpen } = props;
  return(
    <>
      {isOpen?
        (<button onClick={onClickNavBarSwitch} className="sm:top-2 sm:right-3 text-7xl text-red-400 fixed top-0 right-1 h-10 w-14 block ml-4">
          Ã—
        </button>
        ):
        (<button onClick={onClickNavBarSwitch} className="sm:top-7 fixed top-6 right-3 h-10 w-14 block ml-4">
          <div className="absolute sm:w-14 w-10 border-red-400 border-t-4 inline-block top-0 right-0"></div>
          <div className="absolute sm:w-14 w-10 border-red-400 border-t-4 inline-block top-4 right-0"></div>
          <div className="absolute sm:w-14 w-10 border-red-400 border-t-4 inline-block top-8 right-0"></div>
        </button>)
      }
    </>
  );
});