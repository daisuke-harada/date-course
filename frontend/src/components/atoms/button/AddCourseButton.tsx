import { memo, VFC } from "react";
import { useNavigate } from "react-router-dom";
import { useRecoilState, useRecoilValue } from "recoil";
import { managementCourseState } from "store/managementCourse";
import { currentUserState } from "store/session";
import tw from "tailwind-styled-components";
import { AddressAndDateSpotJoinData } from "types/dateSpots/response";
import { BaseButton } from "./BaseButton";

type Props = {
  addressAndDateSpot: AddressAndDateSpotJoinData
}

const ButtonParentDiv = tw.div`m-5`;

export const AddCourseButton: VFC<Props> = memo((props) => {
  const { addressAndDateSpot } = props;
  const navigate = useNavigate();
  const getCurrentUser = useRecoilValue(currentUserState);
  const [managementCourses, setManagementCourses] = useRecoilState(managementCourseState({userId: getCurrentUser.user.id}));
  const onClickAddCourseAction = () => {
    if(managementCourses.userId === 0){
      setManagementCourses({userId: getCurrentUser.user.id, courseDuringSpots: [addressAndDateSpot]});
      navigate('/managementCourses/createCourse');
    } else if(managementCourses.courseDuringSpots.some(spot => spot.dateSpot.id === addressAndDateSpot.dateSpot.id)){
      navigate('./', {state: {message: 'このスポットはすでに選択されています', type: 'error-message', condition: true}});
    } else {
      const dateCourseIdAndNames = managementCourses.courseDuringSpots.slice();
      dateCourseIdAndNames.push(addressAndDateSpot);
      setManagementCourses({userId: getCurrentUser.user.id, courseDuringSpots: dateCourseIdAndNames});
      navigate('/managementCourses/createCourse');
    };
  };
  return(
    <ButtonParentDiv>
       <BaseButton dataE2e={`courseAddButtonId-${addressAndDateSpot.dateSpot.id}`} onClickEvent={onClickAddCourseAction}>コースに追加</BaseButton>
    </ButtonParentDiv>
  );
});
