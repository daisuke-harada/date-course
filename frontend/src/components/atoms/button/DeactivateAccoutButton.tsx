import { client } from "lib/api/client";
import { memo, VFC } from "react";
import { useNavigate } from "react-router-dom";
import { useRecoilState } from "recoil";
import tw from 'tailwind-styled-components';

import { currentUserState, loginStatusState } from "store/session";

import { DangerButton } from "./DangerButton";

const ButtonParentDiv = tw.div`text-center p-1 mx-6 my-4`;

export const DeactivateAccoutButton: VFC = memo(() => {
  const [loginStatus, setLoginStatus] = useRecoilState(loginStatusState);
  const [currentUser, setCurrentUser] = useRecoilState(currentUserState);
  const navigate = useNavigate();

  const onCLickDeactivateAccountAction: React.MouseEventHandler<HTMLButtonElement> = (e) => {
    client.delete(`users/${currentUser.user.id}`).then(response => {
      response.data.status === 'delete' && navigate('/', {state: {message: '退会しました', type: 'success-message', condition: true}} );
      response.data.status === 'delete' && setLoginStatus({status: false}) && setCurrentUser({user:{}});
    });
  };

  return(
    <>
      { loginStatus.status === true
        &&
        <ButtonParentDiv>
          <DangerButton onClickEvent={onCLickDeactivateAccountAction}>退会</DangerButton>
        </ButtonParentDiv>
      }
    </>
  );

});