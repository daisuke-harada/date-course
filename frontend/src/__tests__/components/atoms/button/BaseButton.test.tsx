import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BaseButton } from 'components/atoms/button/BaseButton'

describe('BaseButtonコンポーネントのテスト', () => {
  test('ボタン名が表示される', () => {
    render(
      <BaseButton>
        ボタン
      </BaseButton>
    );
    expect(screen.getByText('ボタン')).toBeInTheDocument();
  });

  test('ボタンをクリックする', () => {
    const mockOnClick = jest.fn();
    render(
      <BaseButton onClickEvent={mockOnClick}>
        ボタン
      </BaseButton>
    );

    // buttonをクリックする
    fireEvent.click(screen.getByRole('button'));

    // mock関数がクリックされて一度呼び出されたことを示す。
    expect(mockOnClick).toHaveBeenCalledTimes(1);

  });
});
