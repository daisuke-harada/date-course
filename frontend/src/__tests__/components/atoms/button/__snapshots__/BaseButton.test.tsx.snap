// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButtonコンポーネントのテスト ボタン名が表示される 1`] = `ShallowWrapper {}`;
